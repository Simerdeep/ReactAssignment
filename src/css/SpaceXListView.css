.card {
    box-shadow: 0 4px 8px 0 rgba(0,0,0,0.2);
    transition: 0.3s;
    background: white;
    margin: 0 8px 8px 8px;
}

.missionTitle{
    font-weight: 700;
    color: #660099;
}

.spacexValue{
    color:#000099;
    margin-left: 5px;
}

.spaceXimage{
    background: #F5F5F5;
    width: 100%;
    height: auto;
}

.spaceXimagebox{
    background: #F5F5F5;
    min-height: 200px;
}



.spaceXView {
    display: flex;
    flex-wrap: wrap;
}