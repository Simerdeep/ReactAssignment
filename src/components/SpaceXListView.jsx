//@flow

import type {Element} from "React";
import React from "react";
import { spacexConfig } from "./spaceXHelper";
import "../css/SpaceXListView.css";
import { NO_DATA_AVAILABLE } from "../AppLabels"

type Props = {
    spaceXListData: Array<Object>
};

/**
 * Component to show the spacex list
 *
 * @param {Props} {
 *     spaceXListData
 * }
 * @returns {Element<"div">}
 */
const SpaceXListView = ({
    spaceXListData
}: Props): Element<"div"> => {


    /**
     * To render the nested view if required ( Required in case of mission ids)
     *
     * @param {*} data data to show in nested view
     * @param {*} config config to check for nested flag
     * @returns Element
     */  
    const getNestedView = (data,config) => {

        if(config.isNested){

            return (
                <ul>
                {
                    data[config.accessor].map((value,index) => <li key={index}>{value}</li>)
                }
                </ul>
            );

        }

        return null;

    }

    /**
     * To get the field value to show in the dom
     *
     * @param {*} data data to extract field from
     * @param {*} config config to extract acccessor from 
     * @returns value to be shown in dom
     */
    const getFieldValue = (data, config) => {

        const { accessor, isNested } = config;

        let value = "";

        if(typeof(accessor) === "function")
            value = accessor(data);
        else
            value = data[accessor];

        if(typeof(value) === "boolean")
            value = value.toString();

        if(!value || isNested)
            value = "";
        
        return value;

    }

    return (
        <div className="spaceXView">

            {

                spaceXListData.length > 0 ?
                <>
                {
                    spaceXListData.map((data,key) => {
                        
                        return (
                            <div className="card col-3" key={key}>

                                <div className="spaceXimagebox">
                                    <img src={data.links.mission_patch} alt="SpaceX" className="spaceXimage"/>
                                </div>
                                {
                                    spacexConfig.map((config,key) => {
                                        
                                        return(
                                            <div key={key}>
                                                {
                                                    config.heading ?
                                                    <p>
                                                        <span className="font-weight-bold">{config.heading}</span> :  
                                                        <span className="spacexValue">
                                                            { getFieldValue(data,config)}
                                                        </span>
                                                        
                                                    </p>
                                                    : <p className={config.customClass}>{getFieldValue(data,config)}</p>
                                                }
                                              
                                                { getNestedView(data,config) }
                                            </div>
                                            
                                        )
                                    })
                                }
        
                            </div>
        
                        )
                       
                    })
                }
                </>
                : <h1>{NO_DATA_AVAILABLE}</h1>
            }
        
        </div>
    );

}

export default SpaceXListView;