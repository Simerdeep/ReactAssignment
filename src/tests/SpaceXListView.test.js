import React from "react";
import { mount } from "enzyme";
import SpaceXListView from "../components/SpaceXListView";

describe("SpaceXListView test cases",() => {

    const spaceXListViewWrapper = mount(<SpaceXListView spaceXListData={[]}/>);

    it("should render the snapshot for spacexlist view", () => {
        expect(spaceXListViewWrapper).toMatchSnapshot();
    });

    it("should render card view when spacexlistdata is there", () => {

        const spaceXListData = [{
            mission_name: "1234567",
            flight_number: "123a",
            mission_id: ["123"],
            launch_year: "123",
            launch_success: false,
            rocket: null,
            links:{
                mission_patch: "23456"
            }
        }];

        spaceXListViewWrapper.setProps({
            spaceXListData
        })

        expect(spaceXListViewWrapper.find(".card").length).toBe(1);
    });

})
