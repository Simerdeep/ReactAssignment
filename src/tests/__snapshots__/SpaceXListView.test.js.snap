// Jest Snapshot v1, https://goo.gl/fbAQLP

exports["SpaceXListView test cases should render the snapshot for spacexlist view 1"] = "ReactWrapper {}";
